src/engine/Server.cpp

```diff
@@ -368,57 +422,88 @@ CPP_template_2(typename RequestT, typename ResponseT)(
-        co_return co_await processQuery(parameters, std::move(parsedOperation),
-                                        requestTimer, cancellationHandle, qec,
-                                        std::move(request), send,
-                                        timeLimit.value());
-      } else {
-        static_assert(std::is_same_v<Operation, Update>);
-        co_return co_await processUpdate(
-            std::move(parsedOperation), requestTimer, cancellationHandle, qec,
-            std::move(request), send, timeLimit.value());
-      }
-    } else {
+  auto visitOperation =
+      [&checkParameter, &accessTokenOk, &request, &send, &parameters,
+       &requestTimer,
+       this](ParsedQuery parsedOperation, std::string operationName,
+             std::function<bool(const ParsedQuery&)> expectedOperation,
+             const std::string msg) -> Awaitable<void> {
+    auto timeLimit = co_await verifyUserSubmittedQueryTimeout(
+        checkParameter("timeout", std::nullopt), accessTokenOk, request, send);
+    if (!timeLimit.has_value()) {
       // If the optional is empty, this indicates an error response has been
       // sent to the client already. We can stop here.
       co_return;
     }
+    ad_utility::websocket::MessageSender messageSender = createMessageSender(
+        queryHub_, request, parsedOperation._originalString);
+
+    auto [qec, cancellationHandle, cancelTimeoutOnDestruction] =
+        prepareOperation(operationName, parsedOperation._originalString,
+                         messageSender, parameters, timeLimit.value());
+    if (!expectedOperation(parsedOperation)) {
+      throw std::runtime_error(
+          absl::StrCat(msg, parsedOperation._originalString));
+    }
+    if (parsedOperation.hasUpdateClause()) {
+      co_return co_await processUpdate(
+          std::move(parsedOperation), requestTimer, cancellationHandle, qec,
+          std::move(request), send, timeLimit.value());
+    } else {
+      AD_CORRECTNESS_CHECK(parsedOperation.hasSelectClause() ||
+                           parsedOperation.hasAskClause() ||
+                           parsedOperation.hasConstructClause());
+      co_return co_await processQuery(
+          parameters, std::move(parsedOperation), requestTimer,
+          cancellationHandle, qec, std::move(request), send, timeLimit.value());
+    }
   };
-  auto visitQuery = [&visitOperation](const Query& query) -> Awaitable<void> {
+  auto parseAndAddDatasets = [](std::string operation,
+                                const std::vector<DatasetClause>& datasets) {
+    auto parsedOperation = SparqlParser::parseQuery(std::move(operation));
+    // SPARQL Protocol 2.1.4 specifies that the dataset from the query
+    // parameters overrides the dataset from the query itself.
+    if (!datasets.empty()) {
+      parsedOperation.datasetClauses_ =
+          parsedQuery::DatasetClauses::fromClauses(datasets);
+    }
+    return parsedOperation;
+  };
```

Comment:
Lambdas without any captures never have to be local (can be just static functions or in the anonymous namespace outside the class, or even member static member functions, which allows for better testing).